### PR TITLE
Add support for built-in functions

### PR DESCRIPTION
**Pull request by dbanck**
This PR adds the function signatures for all previous Terraform versions, starting with 0.12. For versions >= 1.4, we use the new `terraform metadata functions -json` command to generate the signatures.

The changes are currently a no-op for users. However, we will follow up with modifications to the language server to obtain the functions for a specific Terraform version using `FunctionsForVersion`.

I've created [a list of all Terraform functions](https://docs.google.com/spreadsheets/d/1P9Kk--fxhT9E4NilalU3A3oEehYcHb3yQH8Wn7cIUG8/edit?pli=1#gid=0) with their introduction versions for a better overview and used that as a source for all the signatures.

## TODO

- [x] Bump terraform-exec version after https://github.com/hashicorp/terraform-exec/pull/358 is merged
    - [x] Pin to new terraform-exec release
- [x] Bump hcl-lang version after https://github.com/hashicorp/hcl-lang/pull/207 is merged

---

* Part of https://github.com/hashicorp/terraform-ls/issues/37

### CODE REVIEW

File: internal/funcs/0.12/base_functions.go
Comment (radeksimko) [inline internal/funcs/0.12/base_functions.go:59]: I'm not sure where this came from and whether it affects other places in Core as well, but I see some HTML escaping in effect (`&amp;#...`) which probably shouldn't be.

The context in which we need to escape the (Markdown) description text would be single-line Go string, i.e. `"` would need to be `\"` but that's about it. Single quotes and probably most other characters can stay as-is.

I don't know if there's any escaping function we could use in this context in the generator, here I suppose it's just a simple matter of manual Find+Replace.

File: internal/funcs/generated/README.md
Comment (radeksimko) [inline internal/funcs/generated/README.md:5]:
```suggestion
It is intended to run whenever HashiCorp releases a new Terraform version. If the Terraform version contains updated function signatures, it generates a new file for that version. When no changes are detected, one should commit the version bump in `gen/gen.go`.
```

File: internal/funcs/generated/gen/gen.go
Comment (radeksimko) [inline internal/funcs/generated/gen/gen.go:156]: Nitpick: I know this is a script which will be executed by a maintainer, but it's still a good practice for any downstream functions to _return_ error and only the top level to actually deal with surfacing it somehow.

This enables testability and reusability and helps readability (it's easier to understand the flow of the program as a whole), even though admittedly we probably won't be testing or reusing this code any time soon.

File: internal/funcs/generated/gen/gen.go
Comment (radeksimko) [inline internal/funcs/generated/gen/gen.go:189]: Again - not sure if the HTML escape characters come already in the JSON output - which would be something to address upstream, but either way we should make sure that double quotes are escaped correctly here too.

File: schema/functions.go
Comment (radeksimko) [inline schema/functions.go:50]: Given this should never happen, I'd consider not exporting this error - since we should never be catching it in LS.

We could go as far as `panic`ing here, but it's probably still better not to, since it's a library and as such it should not decide how errors are surfaced. LS _could_ however turn that error into `panic`.

File: schema/functions.go
Comment (radeksimko) [inline schema/functions.go:18]: I _think_ the `semVer` predates the [`Core()`](https://pkg.go.dev/github.com/hashicorp/go-version#Version.Core) method upstream, which could now be used directly and we could get rid of `semVer()`.

**Review by radeksimko — COMMENTED** [review]
I'm assuming this is now expected, until 1.4.0 final actually lands, just surfacing it here for posterity:

```
2023/02/16 12:39:16 ensuring terraform is installed
2023/02/16 12:39:17 failed to obtain product version from "https://releases.hashicorp.com/terraform/1.4.0/index.json": 404 Not Found
exit status 1
internal/funcs/generated/doc.go:4: running "go": exit status 1
```

I left some comments in-line, mostly minor, but the escaping one should certainly be resolved before we merge it, especially if it affects upstream as well (i.e. the JSON TF Core produces is HTML escaped).

File: internal/funcs/0.12/base_functions.go
Comment (dbanck) [reply internal/funcs/0.12/base_functions.go]: Good catch! I fixed it in Terraform in https://github.com/hashicorp/terraform/pull/32710 and replaced all occurrences here.

File: internal/funcs/generated/gen/gen.go
Comment (dbanck) [reply internal/funcs/generated/gen/gen.go]: Oh, missed updating the error handling for the two templating functions. Thanks for pointing it out.

File: schema/functions.go
Comment (dbanck) [reply schema/functions.go]: Agreed — replaced it with a simple error string which we can turn into a `panic` upstream.

**Review by radeksimko — APPROVED** [review]
LGTM :shipit:

**Pull request merged**
